crates/resolver/src/name_server/connection_provider.rs

```diff
@@ -94,6 +88,28 @@ pub trait RuntimeProvider: Clone + Send + Sync + Unpin + 'static {
         local_addr: SocketAddr,
         server_addr: SocketAddr,
     ) -> Pin<Box<dyn Send + Future<Output = io::Result<Self::Udp>>>>;
+
+    /// Yields an object that knows how to bind a QUIC socket.
+    //
+    // Use some indirection here to avoid exposing the `quinn` crate in the public API
+    // even for runtimes that might not (want) to provide QUIC support.
+    fn quic_binder(&self) -> Option<&dyn QuicSocketBinder> {
+        None
+    }
+}
+
+/// Noop trait for when the `quinn` dependency is not available.
+#[cfg(not(any(feature = "dns-over-quic", feature = "dns-over-h3")))]
+pub trait QuicSocketBinder {}
+
+#[cfg(any(feature = "dns-over-quic", feature = "dns-over-h3"))]
+pub trait QuicSocketBinder {
+    /// Create a UDP socket for QUIC usage.
+    fn bind_quic(
```

Comment:
This feels very similar to the previous Future that was returned before that performs the binding. Could we make this generic across all the ConnectionProvider impls (future change).

Reply:
Not sure what you mean exactly?

Reply:
I think it's fine right now, I was trying to think if there was a way to abstract out this binder, but we can look at that in a future change to bring some of this back together.